File: snippets/fsharp/System/ValueType/Overview/example1.fs
// <Snippet1>
open System
open System.Numerics

module Utility =
    type NumericRelationship =
        | GreaterThan = 1 
        | EqualTo = 0
        | LessThan = -1
   
    let isInteger (value: ValueType) =
        match value with
        | :? sbyte | :? int16 | :? int32 | :? int64 
        | :? byte | :? uint16  | :? uint32 
        | :? uint64 | :? bigint -> true
        | _ -> false

    let isFloat (value: ValueType) = 
        match value with
        | :? float32 | :? float | :? decimal -> true
        | _ -> false

    let tryToBigInt (value: ValueType) =
        match value with
        | :? sbyte as v -> bigint v |> Some
        | :? int16 as v -> bigint v |> Some
        | :? int32 as v -> bigint v |> Some
        | :? int64 as v -> bigint v |> Some
        | :? byte as v -> bigint v |> Some
        | :? uint16 as v -> bigint v |> Some
        | :? uint32 as v -> bigint v |> Some
        | :? uint64 as v -> bigint v |> Some
        | :? float32 as v -> bigint v |> Some
        | _ -> None

    let isNumeric (value: ValueType) =
        isInteger value || isFloat value

    let compare (value1: ValueType) (value2: ValueType) =
        if isNumeric value1 |> not then
            invalidArg "value1" "value1 is not a number."
        elif isNumeric value2 |> not then
            invalidArg "value2" "value2 is not a number."

        // Use BigInteger as common integral type
        match tryToBigInt value1, tryToBigInt value2 with
        | Some bigint1, Some bigint2 ->
            BigInteger.Compare(bigint1, bigint2) |> enum<NumericRelationship>

        // At least one value is floating point use Double.
        | _ ->
            let dbl1 = 
                try
                    Convert.ToDouble value1
                with
                | :? OverflowException ->
                    printfn "value1 is outside the range of a Double."
                    0.
                | _ -> 0.

            let dbl2 = 
                try
                    Convert.ToDouble value2
                with
                | :? OverflowException ->
                    printfn "value2 is outside the range of a Double."
                    0.
                | _ -> 0.
                
            dbl1.CompareTo dbl2 |> enum<NumericRelationship>
// </Snippet1>

// <Snippet2> 
printfn $"{Utility.isNumeric 12}"
printfn $"{Utility.isNumeric true}"
printfn $"{Utility.isNumeric 'c'}"
printfn $"{Utility.isNumeric (DateTime(2012, 1, 1))}"
printfn $"{Utility.isInteger 12.2}"
printfn $"{Utility.isInteger 123456789}"
printfn $"{Utility.isFloat true}"
printfn $"{Utility.isFloat 12.2}"
printfn $"{Utility.isFloat 12}"
printfn $"{12.1} {Utility.compare 12.1 12} {12}"
// The example displays the following output:
//       True
//       False
//       False
//       False
//       False
//       True
//       False
//       True
//       False
//       12.1 GreaterThan 12
// </Snippet2>

File: snippets/csharp/System/ValueType/Overview/example1.cs
// <Snippet1>
using System;
using System.Numerics;

public class Utility
{
    public enum NumericRelationship
    {
        GreaterThan = 1,
        EqualTo = 0,
        LessThan = -1
    };

    public static NumericRelationship Compare(ValueType value1, ValueType value2)
    {
        if (!IsNumeric(value1))
            throw new ArgumentException("value1 is not a number.");
        else if (!IsNumeric(value2))
            throw new ArgumentException("value2 is not a number.");

        // Use BigInteger as common integral type
        if (IsInteger(value1) && IsInteger(value2))
        {
            BigInteger bigint1 = (BigInteger)value1;
            BigInteger bigint2 = (BigInteger)value2;
            return (NumericRelationship)BigInteger.Compare(bigint1, bigint2);
        }
        // At least one value is floating point; use Double.
        else
        {
            double dbl1 = 0;
            double dbl2 = 0;
            try
            {
                dbl1 = Convert.ToDouble(value1);
            }
            catch (OverflowException)
            {
                Console.WriteLine("value1 is outside the range of a Double.");
            }
            try
            {
                dbl2 = Convert.ToDouble(value2);
            }
            catch (OverflowException)
            {
                Console.WriteLine("value2 is outside the range of a Double.");
            }
            return (NumericRelationship)dbl1.CompareTo(dbl2);
        }
    }

    public static bool IsInteger(ValueType value) => (value is sbyte || value is short || value is int
               || value is long || value is byte || value is ushort
               || value is uint || value is ulong
               || value is BigInteger);

    public static bool IsFloat(ValueType value) => (value is float || value is double || value is decimal);

    public static bool IsNumeric(ValueType value) => (value is byte ||
               value is short ||
               value is int ||
               value is long ||
               value is sbyte ||
               value is ushort ||
               value is uint ||
               value is ulong ||
               value is BigInteger ||
               value is decimal ||
               value is double ||
               value is float);
}
// </Snippet1>

// <Snippet2>
public class Example
{
    public static void Main()
    {
        Console.WriteLine(Utility.IsNumeric(12));
        Console.WriteLine(Utility.IsNumeric(true));
        Console.WriteLine(Utility.IsNumeric('c'));
        Console.WriteLine(Utility.IsNumeric(new DateTime(2012, 1, 1)));
        Console.WriteLine(Utility.IsInteger(12.2));
        Console.WriteLine(Utility.IsInteger(123456789));
        Console.WriteLine(Utility.IsFloat(true));
        Console.WriteLine(Utility.IsFloat(12.2));
        Console.WriteLine(Utility.IsFloat(12));
        Console.WriteLine($"{12.1} {Utility.Compare(12.1, 12)} {12}");
    }
}
// The example displays the following output:
//       True
//       False
//       False
//       False
//       False
//       True
//       False
//       True
//       False
//       12.1 GreaterThan 12
// </Snippet2>
